hail/src/main/scala/is/hail/io/RowStore.scala

```diff
@@ -986,343 +935,224 @@ object EmitPackDecoder {
-            val r2 = allocState(s"${name}.missing")
-            mainCode.append(s"  auto ngot = this->decode_bytes(&${miss}[${idx}], ${t.nMissingBytes}-${idx});\n")
-            mainCode.append(s"  if (ngot <= 0) { s = ${r2}; goto pull; }\n")
-            mainCode.append(s"  ${idx} += ngot;\n")
-            mainCode.append(s"}\n")
-          }
-          var fieldIdx = 0
-          while (fieldIdx < t.fields.length) {
-            val field = t.fields(fieldIdx)
-            val wantIdx = fieldToWantIdx(fieldIdx)
-            val fieldSkip = skip || (wantIdx < 0)
-            val fieldType = t.types(fieldIdx)
-            val wantType = if (fieldSkip) fieldType else wantStruct.types(wantIdx)
-            val wantOffset = if (fieldSkip) -1 else wantStruct.byteOffsets(wantIdx)
-            if (!t.fieldRequired(fieldIdx)) {
-              val m = t.missingIdx(fieldIdx)
-              mainCode.append(s"if (!is_missing(${miss}, ${m})) {\n")
-              if (!fieldSkip) {
-                if (shuffleMissingBits) {
-                  val mbit = wantStruct.missingIdx(wantIdx)
-                  mainCode.append(s"  ${addr}[${mbit>>3}] &= ~(1<<${mbit&0x7});\n")
-                }
-              }
-              mainCode.append(s"  // ${name}.${field.name} fieldSkip ${fieldSkip} ${fieldType}\n")
-              scan(depth+1, s"${name}.${field.name}", fieldType, wantType, fieldSkip, baseAddr, off+wantOffset)
-              mainCode.append(s"}\n")
-            } else {
-              scan(depth+1, s"${name}.${field.name}", fieldType, wantType, fieldSkip, baseAddr, off+wantOffset)
-            }
-            fieldIdx += 1
-          }
+  def decodeBaseStruct(t: PBaseStruct, rt: PBaseStruct, input_buf_ptr: cxx.Expression, region: cxx.Expression, off: cxx.Expression): cxx.Code = {
+    val ssb = new cxx.StagedBaseStructBuilder(rt, off)
+    def wrapMissing(missingBytes: cxx.Code, tidx: Int)(processField: cxx.Code): cxx.Code = {
+      if (!t.fieldRequired(tidx)) {
+        s"""if (!${t.cxxIsFieldMissing(missingBytes, tidx)}) {
+           |  $processField
+           |}""".stripMargin
+      } else processField
+    }
+    if (t.size == rt.size) {
+      assert((t.isInstanceOf[PTuple] && rt.isInstanceOf[PTuple]) || (t.asInstanceOf[PStruct].fieldNames sameElements t.asInstanceOf[PStruct].fieldNames))
+      val decodeFields = Array.tabulate[cxx.Code](rt.size) { idx =>
+        wrapMissing(s"$off", idx)(ssb.addField(idx, foff => decode(t.types(idx), rt.types(idx), input_buf_ptr, region, cxx.Expression(foff))))
+      }.mkString("\n")
+      if (t.nMissingBytes > 0) {
+        s"""$input_buf_ptr->read_bytes($off, ${ t.nMissingBytes });
+           |$decodeFields""".stripMargin
+      } else
+        decodeFields
+    } else {
+      val names = t.asInstanceOf[PStruct].fieldNames
+      val rnames = rt.asInstanceOf[PStruct].fieldNames
+      val missing_bytes = cxx.ArrayVariable(s"missing", "char", s"${t.nMissingBytes}")
```

Comment:
I might call this `t_missing_bytes` or something to distinguish since we have two sets of missing bits flying around.